File: ios/RevKit/Sources/RevKit/Game/TileScoring.swift
import CoreLocation
import Foundation
import SwiftyH3

/// computes the per-tile claim score (time-weighted average speed while moving) from
/// a (filtered) GPS path per H3 res-10 cell. Mirrors the server math in REF:docs/game-design.md
/// §Score Metric
///
/// score(tile) = Σ speed·time spent moving inside the tile / Σ time spent moving inside the tile.
/// Segment speed is the device's Doppler reading (`CLLocation.speed`) when valid, falling back to
/// the position-derived speed (Δdistance / Δtime) otherwise
///
/// Returned in **mph**; Score Metric v2 converts this raw speed into claim strength.
public enum TileScoring {
    /// 1 m/s expressed in mph
    static let mphPerMetersPerSecond = 2.2369362920544

    /// segments slower than this count as "stopped"
    static let stoppedSpeedMetersPerSecond = 2.0 / mphPerMetersPerSecond

    /// max time between consecutive fixes before a segment is treated as a tracking gap rather than
    /// real travel
    public static let maxSegmentGap: TimeInterval = 10

    /// fastest a segment can imply (m/s) before it's treated as a GPS teleport rather than
    /// real travel
    static let maxSegmentSpeedMetersPerSecond = 134.0

    /// target length (m) of each sub-step a segment is broken into for tile attribution. small
    /// relative to a res-10 hex (~130 m wide) so a segment is credited to each tile in proportion
    /// to how much of it actually lies inside that tile.
    static let subStepMeters = 8.0

    /// cap on sub-steps per segment so a long GPS-gap segment can't blow up the loop.
    static let maxSubSteps = 64

    /// minimum distance (m) actually travelled inside a tile before it earns a score. a tile you
    /// merely clip the edge of has too little in-tile travel to trust - its score would be a single
    /// noisy GPS segment with nothing to average against, which read far too high. below this floor
    /// the tile simply isn't scored (it may still be weakly claimed live at score 0).
    static let minTileDistanceMeters = 12.0

    /// pads the Doppler-vs-position agreement envelope (m/s) for timestamp jitter and
    /// chord-shortening through curves
    static let dopplerAgreementSlack = 3.0

    public static func perTileScores(for samples: [GPSSample]) -> [UInt64: Double] {
        guard samples.count > 1 else { return [:] }

        var distanceByTile: [UInt64: Double] = [:]
        var movingTimeByTile: [UInt64: Double] = [:]
        // scored speed integrated over time (Σ speed·time, in metres). dividing by moving time gives
        // the tile's time-weighted average speed. kept separate from the raw geometric distance
        // because the scored speed comes from the device's Doppler reading, not the position deltas.
        var speedTimeByTile: [UInt64: Double] = [:]

        for (a, b) in zip(samples, samples.dropFirst()) {
            let dt = b.timestamp.timeIntervalSince(a.timestamp)
            guard dt > 0, dt <= maxSegmentGap else { continue }

            let distance = GPSOutlierFilter.distanceMeters(a, b)
            let segmentSpeed = distance / dt
            guard segmentSpeed >= stoppedSpeedMetersPerSecond else { continue }
            // a GPS teleport: don't interpolate a line between two far-apart fixes
            guard segmentSpeed <= maxSegmentSpeedMetersPerSecond else { continue }

            // position error only ever adds apparent path length, so Doppler below the
            // position-implied speed is the multipath case Doppler exists for. Doppler above
            // what the positions plus their accuracy budget support is a chip glitch: drop the
            // segment rather than trust either signal. bounds glitches without capping
            // legitimate speed, which always comes with matching position deltas
            var scoreSpeed = segmentSpeed
            if a.speed >= 0, b.speed >= 0 {
                let dopplerSpeed = (a.speed + b.speed) / 2
                let tolerance = (max(a.accuracy, 0) + max(b.accuracy, 0)) / dt + dopplerAgreementSlack
                guard dopplerSpeed <= segmentSpeed + tolerance else { continue }
                scoreSpeed = dopplerSpeed
            }

            // walk the segment in small steps and attribute each step's distance/time to the tile
            // it falls in. this keeps a barely-clipped tile from inheriting a whole segment's worth
            // of travel that mostly happened in a neighbor (which inflates its score)
            let steps = max(1, min(maxSubSteps, Int((distance / subStepMeters).rounded(.up))))
            let stepDistance = distance / Double(steps)
            let stepTime = dt / Double(steps)
            for i in 0..<steps {
                let fraction = (Double(i) + 0.5) / Double(steps)
                let lat = a.lat + (b.lat - a.lat) * fraction
                let lng = a.lng + (b.lng - a.lng) * fraction
                guard let tile = cell(lat: lat, lng: lng) else { continue }
                distanceByTile[tile, default: 0] += stepDistance
                movingTimeByTile[tile, default: 0] += stepTime
                speedTimeByTile[tile, default: 0] += scoreSpeed * stepTime
            }
        }

        var scores: [UInt64: Double] = [:]
        for (tile, distance) in distanceByTile {
            guard distance >= minTileDistanceMeters else { continue }
            guard let movingTime = movingTimeByTile[tile], movingTime > 0 else { continue }
            scores[tile] = ((speedTimeByTile[tile] ?? 0) / movingTime) * mphPerMetersPerSecond
        }
        return scores
    }

    /// total distance moved and time spent moving across a (filtered) path
    public static func movementStats(for samples: [GPSSample]) -> (distanceMeters: Double, movingTime: TimeInterval) {
        guard samples.count > 1 else { return (0, 0) }

        var totalDistance = 0.0
        var totalMovingTime = 0.0
        for (a, b) in zip(samples, samples.dropFirst()) {
            let dt = b.timestamp.timeIntervalSince(a.timestamp)
            guard dt > 0, dt <= maxSegmentGap else { continue }

            let distance = GPSOutlierFilter.distanceMeters(a, b)
            let segmentSpeed = distance / dt
            guard segmentSpeed >= stoppedSpeedMetersPerSecond else { continue }
            guard segmentSpeed <= maxSegmentSpeedMetersPerSecond else { continue }

            totalDistance += distance
            totalMovingTime += dt
        }
        return (totalDistance, totalMovingTime)
    }

    /// ordered and de-duplicated list of res-10 cells the path passes through (first-seen order)
    /// to be used by the live-claim path to know which tiles a drive has entered
    public static func tilesCrossed(for samples: [GPSSample]) -> [UInt64] {
        var seen: Set<UInt64> = []
        var ordered: [UInt64] = []
        for sample in samples {
            guard let tile = cell(lat: sample.lat, lng: sample.lng) else { continue }
            if seen.insert(tile).inserted { ordered.append(tile) }
        }
        return ordered
    }

    /// split a raw path into continuous breadcrumb segments, breaking wherever the gap between
    /// fixes exceeds `maxSegmentGap` (the app was suspended and we don't know the route taken). the
    /// map draws one polyline per segment so a gap stays a gap instead of a false straight line.
    public static func segmentedPath(_ path: [GPSSample]) -> [[CLLocationCoordinate2D]] {
        var segments: [[CLLocationCoordinate2D]] = []
        var current: [CLLocationCoordinate2D] = []
        var previous: Date?
        for sample in path {
            if let previous, sample.timestamp.timeIntervalSince(previous) > maxSegmentGap {
                segments.append(current)
                current = []
            }
            current.append(CLLocationCoordinate2D(latitude: sample.lat, longitude: sample.lng))
            previous = sample.timestamp
        }
        if !current.isEmpty { segments.append(current) }
        return segments
    }

    static func cell(lat: Double, lng: Double) -> UInt64? {
        let coordinate = CLLocationCoordinate2D(latitude: lat, longitude: lng)
        return H3Grid.cell(for: coordinate)?.id
    }
}

File: backend/internal/game/tilescoring.go
package game

import (
	"math"

	"github.com/camden-git/rev-monorepo/backend/internal/h3util"
)

// Point is one GPS fix's coordinate (the lat/lng of a drives.raw_path entry)
type Point struct {
	Lat float64
	Lng float64
}

// Tile-scoring constants, mirrored from RevKit TileScoring
const (
	// MphPerMetersPerSecond is 1 m/s expressed in mph
	MphPerMetersPerSecond = 2.2369362920544
	// stoppedSpeedMetersPerSecond: segments slower than this count as "stopped"
	stoppedSpeedMetersPerSecond = 2.0 / MphPerMetersPerSecond
	// MaxSegmentGap is the max time (s) between fixes before a segment is treated
	// as a tracking gap rather than real travel
	MaxSegmentGap = 10.0
	// maxSegmentSpeed is the fastest a segment can imply (m/s, ~200 mph) before
	// it's treated as a GPS teleport rather than real travel
	maxSegmentSpeed = 134
	// subStepMeters is the target length each segment is broken into for tile
	// attribution (small relative to a ~130 m res-10 hex)
	subStepMeters = 8.0
	// maxSubSteps caps sub-steps per segment so a long GPS-gap segment can't blow
	// up the loop
	maxSubSteps = 64
	// minTileDistanceMeters is the minimum in-tile travel before a tile earns a
	// score; below this the tile is too noisy to trust
	minTileDistanceMeters = 12.0
)

// Cell returns the res-10 cell containing a coordinate
// Port of the `cell(lat:lng:)` helper in RevKit TileScoring
func Cell(lat, lng float64) (uint64, bool) {
	return h3util.LatLngToCell(lat, lng)
}

// PerTileScores computes the per-tile claim score (time-weighted average speed
// while moving, in mph) for each res-10 cell a filtered path crosses. Segment
// speed comes from the device's Doppler reading when valid
func PerTileScores(samples []Sample) map[uint64]float64 {
	if len(samples) <= 1 {
		return map[uint64]float64{}
	}

	distanceByTile := map[uint64]float64{}
	movingTimeByTile := map[uint64]float64{}

	speedTimeByTile := map[uint64]float64{}

	for i := 0; i+1 < len(samples); i++ {
		a, b := samples[i], samples[i+1]
		dt := b.TS.Sub(a.TS).Seconds()
		if dt <= 0 || dt > MaxSegmentGap {
			continue
		}
		distance := distanceMeters(a, b)
		segmentSpeed := distance / dt
		if segmentSpeed < stoppedSpeedMetersPerSecond {
			continue
		}
		if segmentSpeed > maxSegmentSpeed {
			continue // GPS teleport: don't interpolate a line between far-apart fixes
		}

		// score from the device's Doppler speed when both endpoints report a
		// valid reading. Doppler speed is far steadier than differentiating GPS
		// positions, which multipath (e.g. downtown urban canyons) inflates: a
		// gradual position wander only ever adds apparent path length and reads
		// as phantom speed, yet is too smooth to trip the spike/acceleration
		// filters. fall back to the position-derived speed when Doppler is
		// unavailable (negative CoreLocation speed = invalid fix).
		scoreSpeed := segmentSpeed
		if a.Speed >= 0 && b.Speed >= 0 {
			scoreSpeed = (a.Speed + b.Speed) / 2
		}

		// walk the segment in small steps and attribute each step's distance/time
		// to the tile it falls in
		steps := int(math.Ceil(distance / subStepMeters))
		if steps < 1 {
			steps = 1
		}
		if steps > maxSubSteps {
			steps = maxSubSteps
		}
		stepDistance := distance / float64(steps)
		stepTime := dt / float64(steps)
		for s := 0; s < steps; s++ {
			fraction := (float64(s) + 0.5) / float64(steps)
			lat := a.Lat + (b.Lat-a.Lat)*fraction
			lng := a.Lng + (b.Lng-a.Lng)*fraction
			tile, ok := h3util.LatLngToCell(lat, lng)
			if !ok {
				continue
			}
			distanceByTile[tile] += stepDistance
			movingTimeByTile[tile] += stepTime
			speedTimeByTile[tile] += scoreSpeed * stepTime
		}
	}

	scores := map[uint64]float64{}
	for tile, distance := range distanceByTile {
		if distance < minTileDistanceMeters {
			continue
		}
		movingTime := movingTimeByTile[tile]
		if movingTime <= 0 {
			continue
		}
		scores[tile] = (speedTimeByTile[tile] / movingTime) * MphPerMetersPerSecond
	}
	return scores
}

// TilesCrossed returns the ordered, de-duplicated res-10 cells a path passes
// through. port of RevKit TileScoring.tilesCrossed
func TilesCrossed(path []Point) []uint64 {
	seen := map[uint64]bool{}
	ordered := []uint64{}
	for _, p := range path {
		cell, ok := h3util.LatLngToCell(p.Lat, p.Lng)
		if !ok {
			continue
		}
		if !seen[cell] {
			seen[cell] = true
			ordered = append(ordered, cell)
		}
	}
	return ordered
}
